controllers/azurejson_machinepool_controller_test.go

```diff
@@ -210,3 +217,140 @@ func TestAzureJSONPoolReconciler(t *testing.T) {
 		})
 	}
 }
+
+func TestAzureJSONPoolReconcilerUserAssignedIdentities(t *testing.T) {
+	g := NewWithT(t)
+	ctrlr := gomock.NewController(t)
+	defer ctrlr.Finish()
+	req := ctrl.Request{NamespacedName: types.NamespacedName{Name: "fake-machine-pool", Namespace: "fake-ns"}}
+	ctx := context.Background()
+	scheme := runtime.NewScheme()
+
+	err := infrav1exp.AddToScheme(scheme)
+	g.Expect(err).ToNot(HaveOccurred())
+
+	err = expv1.AddToScheme(scheme)
+	g.Expect(err).ToNot(HaveOccurred())
+
+	err = corev1.AddToScheme(scheme)
+	g.Expect(err).ToNot(HaveOccurred())
+
+	err = clusterv1.AddToScheme(scheme)
+	g.Expect(err).ToNot(HaveOccurred())
+
+	err = infrav1.AddToScheme(scheme)
+	g.Expect(err).ToNot(HaveOccurred())
+
+	azureMP := &infrav1exp.AzureMachinePool{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "fake-machine-pool",
+			Namespace: "fake-ns",
+			Labels: map[string]string{
+				clusterv1.ClusterNameLabel: "fake-cluster",
+			},
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: expv1.GroupVersion.Group + "/" + expv1.GroupVersion.Version,
+					Kind:       "MachinePool",
+					Name:       "fake-other-machine-pool",
+					Controller: to.Ptr(true),
+				},
+			},
+		},
+		Spec: infrav1exp.AzureMachinePoolSpec{
+			UserAssignedIdentities: []infrav1.UserAssignedIdentity{
+				{
+					ProviderID: "fake-id",
+				},
+			},
+		},
+	}
+
+	clusterA := &clusterv1.Cluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "fake-cluster",
+			Namespace: "fake-ns",
+		},
+		Spec: clusterv1.ClusterSpec{
+			InfrastructureRef: &corev1.ObjectReference{
+				Kind:      "AzureCluster",
+				Name:      "fake-azure-cluster",
+				Namespace: "fake-ns",
+			},
+		},
+	}
+
+	ownerMP := &expv1.MachinePool{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "fake-other-machine-pool",
+			Namespace: "fake-ns",
+			Labels: map[string]string{
+				clusterv1.ClusterNameLabel: "fake-cluster",
+			},
+		},
+	}
+
+	azureCluster := &infrav1.AzureCluster{
```

Comment:
I think you can use this `getFakeAzureCluster` function that exists here: https://github.com/willie-yao/cluster-api-provider-azure/blob/aks-fleet/controllers/azuremachine_controller_test.go#L529

However, I think there could be a case to be made to create a new file that contains all of these functions for tests, since it's a bit confusing they exist in specific controller tests but can be used across the package. I think this could be done in a follow-up PR since it'll be a bit out of scope for this issue.

Reply:
I would definitely do this but to migrate from this to that may be additional effort that may pollute the addition of this test.  

Also, setting up fixtures to settle all cases where that may be useful may also pollute the repo where that may need to be tested to ensure code coverage (I know that metric isn't bound to quality, but I digress).  

I can create an issue and follow up to call that out as something we would like to have.

Reply:
Sounds good! I think it would be good to leave as-is for the current PR